// @adonisjs/lucid/build/src/Orm/Decorators/index.js
"use strict";
/*
* @adonisjs/lucid
*
* (c) Harminder Virk <virk@adonisjs.com>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.afterPaginate = exports.beforePaginate = exports.afterFetch = exports.beforeFetch = exports.afterFind = exports.beforeFind = exports.afterDelete = exports.beforeDelete = exports.afterUpdate = exports.beforeUpdate = exports.afterCreate = exports.beforeCreate = exports.afterSave = exports.beforeSave = exports.hasManyThrough = exports.manyToMany = exports.hasMany = exports.hasOne = exports.belongsTo = exports.computed = exports.column = void 0;
/// <reference path="../../../adonis-typings/index.ts" />
require("reflect-metadata");
const date_1 = require("./date");
const datetime_1 = require("./datetime");
/**
 * Define property on a model as a column. The decorator needs a
 * proper model class inheriting the base model
 */
exports.column = (options) => {
    return function decorateAsColumn(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addColumn(property, options || {});
    };
};
exports.column.date = date_1.dateColumn;
exports.column.dateTime = datetime_1.dateTimeColumn;
/**
 * Define computed property on a model. The decorator needs a
 * proper model class inheriting the base model
 */
exports.computed = (options) => {
    return function decorateAsComputed(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addComputed(property, options || {});
    };
};
/**
 * Define belongsTo relationship
 */
exports.belongsTo = (relatedModel, relation) => {
    return function decorateAsRelation(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addRelation(property, 'belongsTo', relatedModel, Object.assign({ relatedModel }, relation));
    };
};
/**
 * Define hasOne relationship
 */
exports.hasOne = (relatedModel, relation) => {
    return function decorateAsRelation(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addRelation(property, 'hasOne', relatedModel, Object.assign({ relatedModel }, relation));
    };
};
/**
 * Define hasMany relationship
 */
exports.hasMany = (relatedModel, relation) => {
    return function decorateAsRelation(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addRelation(property, 'hasMany', relatedModel, Object.assign({ relatedModel }, relation));
    };
};
/**
 * Define manyToMany relationship
 */
exports.manyToMany = (relatedModel, relation) => {
    return function decorateAsRelation(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addRelation(property, 'manyToMany', relatedModel, Object.assign({ relatedModel }, relation));
    };
};
/**
 * Define hasManyThrough relationship
 */
exports.hasManyThrough = ([relatedModel, throughModel], relation) => {
    return function decorateAsRelation(target, property) {
        const Model = target.constructor;
        Model.boot();
        Model.$addRelation(property, 'hasManyThrough', relatedModel, Object.assign({ relatedModel, throughModel }, relation));
    };
};
/**
 * Before/After save hook
 */
exports.beforeSave = () => {
    return function decorateAsHook(target, property) {
        target.boot();
        target.before('save', target[property].bind(target));
    };
};
exports.afterSave = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('save', target[property].bind(target));
    };
};
/**
 * Before/After create hook
 */
exports.beforeCreate = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.before('create', target[property].bind(target));
    };
};
exports.afterCreate = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('create', target[property].bind(target));
    };
};
/**
 * Before/After update hook
 */
exports.beforeUpdate = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.before('update', target[property].bind(target));
    };
};
exports.afterUpdate = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('update', target[property].bind(target));
    };
};
/**
 * Before/After delete hook
 */
exports.beforeDelete = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.before('delete', target[property].bind(target));
    };
};
exports.afterDelete = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('delete', target[property].bind(target));
    };
};
/**
 * Before/After find hook
 */
exports.beforeFind = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.before('find', target[property].bind(target));
    };
};
exports.afterFind = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('find', target[property].bind(target));
    };
};
/**
 * Before/After fetchs hook
 */
exports.beforeFetch = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.before('fetch', target[property].bind(target));
    };
};
exports.afterFetch = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('fetch', target[property].bind(target));
    };
};
/**
 * Before/After paginate hook
 */
exports.beforePaginate = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.before('paginate', target[property].bind(target));
    };
};
exports.afterPaginate = () => {
    return function decorateAsColumn(target, property) {
        target.boot();
        target.after('paginate', target[property].bind(target));
    };
};
