// @adonisjs/lucid/build/src/Orm/Relations/BelongsTo/QueryClient.js
"use strict";
/*
 * @adonisjs/lucid
 *
 * (c) Harminder Virk <virk@adonisjs.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.BelongsToQueryClient = void 0;
const QueryBuilder_1 = require("./QueryBuilder");
const utils_1 = require("../../../utils");
/**
 * Query client for executing queries in scope to the belongsTo relationship.
 */
class BelongsToQueryClient {
    constructor(relation, parent, client) {
        this.relation = relation;
        this.parent = parent;
        this.client = client;
    }
    /**
     * Returns value for the foreign key from the related model
     */
    getForeignKeyValue(related, action) {
        return utils_1.getValue(related, this.relation.localKey, this.relation, action);
    }
    /**
     * Generate a query builder instance
     */
    static query(client, relation, rows) {
        const query = new QueryBuilder_1.BelongsToQueryBuilder(client.knexQuery(), client, rows, relation);
        typeof (relation.onQueryHook) === 'function' && relation.onQueryHook(query);
        return query;
    }
    /**
     * Generate a eager query
     */
    static eagerQuery(client, relation, rows) {
        const query = new QueryBuilder_1.BelongsToQueryBuilder(client.knexQuery(), client, rows, relation);
        query.isEagerQuery = true;
        typeof (relation.onQueryHook) === 'function' && relation.onQueryHook(query);
        return query;
    }
    /**
     * Returns instance of query builder
     */
    query() {
        return BelongsToQueryClient.query(this.client, this.relation, this.parent);
    }
    /**
     * Associate the related model with the parent model
     */
    async associate(related) {
        await utils_1.managedTransaction(this.parent.$trx || this.client, async (trx) => {
            related.$trx = trx;
            await related.save();
            this.parent[this.relation.foreignKey] = this.getForeignKeyValue(related, 'associate');
            this.parent.$trx = trx;
            await this.parent.save();
        });
    }
    /**
     * Drop association
     */
    async dissociate() {
        this.parent[this.relation.foreignKey] = null;
        await this.parent.save();
    }
}
exports.BelongsToQueryClient = BelongsToQueryClient;
